File: 2019/day01/python/solution.py
import math
from typing import Union


def calculate_fuel(mass):
    return math.floor(mass / 3) - 2

def calculate_recursive_fuel(mass):
    fuel = calculate_fuel(mass)
    if fuel <= 0:
        return 0
    return fuel + calculate_recursive_fuel(fuel)

def part1(input_data: Union[str, int]):
    if isinstance(input_data, str) and "\n" in input_data:
        masses = [int(line.strip()) for line in input_data.split("\n")]
    elif isinstance(input_data, str):
        masses = [int(input_data.strip())]
    else:
        masses = [input_data]

    total_fuel = sum(calculate_fuel(mass) for mass in masses)
    return str(total_fuel)


def part2(input_data):
    if isinstance(input_data, str) and "\n" in input_data:
        masses = [int(line.strip()) for line in input_data.split("\n")]
    elif isinstance(input_data, str):
        masses = [int(input_data.strip())]
    else:
        masses = [input_data]

    total_fuel = sum(calculate_recursive_fuel(mass) for mass in masses)
    return str(total_fuel)

File: 2019/day01/javascript/solution.js
function calculateFuel(mass) {
  return Math.floor(mass / 3) - 2;
}

function calculateRecursiveFuel(mass) {
  const fuel = Math.floor(mass / 3) - 2;
  if (fuel <= 0) return 0;
  return fuel + calculateRecursiveFuel(fuel);
}

function part1(input) {
  if (input.includes("\n")) {
    return input
      .split("\n")
      .map((a) => calculateFuel(parseInt(a)))
      .reduce((a, b) => a + b, 0)
      .toString();
  }
  return (Math.floor(input / 3) - 2).toString();
}

function part2(input) {
  if (input.includes("\n")) {
    return input
      .split("\n")
      .map((a) => calculateRecursiveFuel(parseInt(a)))
      .reduce((a, b) => a + b, 0)
      .toString();
  }
  return (Math.floor(input / 3) - 2).toString();
}

module.exports = { part1, part2 };
